src/components/cylc/table/deltas.js

```diff
@@ -0,0 +1,196 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import { mergeWith } from 'lodash'
+import Vue from 'vue'
+import WorkflowState from '@/model/WorkflowState.model'
+import { mergeWithCustomizer } from '@/components/cylc/common/merge'
+
+/**
+ * Deltas added.
+ *
+ * @param {DeltasAdded} added
+ * @param {Table} table
+ * @param {Lookup} lookup
+ * @returns {Result}
+ */
+function applyDeltasAdded (added, table, lookup) {
+  const result = {
+    errors: []
+  }
+  if (added.taskProxies) {
+    for (const taskProxy of added.taskProxies) {
+      if (!table[taskProxy.id]) {
+        const tableEntry = {
+          id: taskProxy.id,
+          node: lookup[taskProxy.id],
+          latestJob: {}
+        }
+        Vue.set(table, taskProxy.id, tableEntry)
+      }
+    }
+  }
+  if (added.jobs) {
+    for (const job of added.jobs) {
+      const existingEntry = table[job.firstParent.id]
+      if (existingEntry) {
+        const latestJobSubmitNum = existingEntry.latestJob.submitNum || 0
+        const latestJob = latestJobSubmitNum < job.submitNum
+          ? lookup[job.id]
+          : existingEntry.latestJob
+        Vue.set(existingEntry, 'latestJob', latestJob)
+      }
+    }
+  }
+  return result
+}
+
+/**
+ * Deltas updated.
+ *
+ * @param {DeltasUpdated} updated
+ * @param {Table} table
+ * @param {Lookup} lookup - Not really used since we can fetch the data from the table, and we are not altering the lookup data
+ * @returns {Result}
+ */
+function applyDeltasUpdated (updated, table, lookup) {
+  const result = {
+    errors: []
+  }
+  if (updated.taskProxies) {
+    for (const taskProxy of updated.taskProxies) {
+      if (table[taskProxy.id]) {
+        mergeWith(table[taskProxy.id].node, taskProxy, mergeWithCustomizer)
+      }
+    }
+  }
+  if (updated.jobs) {
+    for (const job of updated.jobs) {
+      if (job.firstParent && job.firstParent.id) {
+        const existingTask = table[job.firstParent.id]
+        if (existingTask && existingTask.latestJob.id === job.id) {
+          mergeWith(existingTask.latestJob, job, mergeWithCustomizer)
+        }
+      }
+    }
+  }
+  return result
+}
+
+/**
+ * Deltas pruned.
+ *
+ * @param {DeltasPruned} pruned
+ * @param {Table} table
+ * @param {Lookup} lookup
+ * @returns {Result}
+ */
+function applyDeltasPruned (pruned, table, lookup) {
+  const result = {
+    errors: []
+  }
+  if (pruned.taskProxies) {
+    pruned.taskProxies.forEach(taskProxyId => {
+      Vue.delete(table, taskProxyId)
+    })
+  }
+  if (pruned.jobs) {
+    pruned.jobs.forEach(jobId => {
+      // TODO: should we use an internal lookup for table too? To replace this loop by a quick O(1) operation
+      //       to fetch the existing job with its ID, and then its existing parent task?
```

Comment:
Looks like we will have to change the table component later, to support showing all the jobs (maybe collapsing too), sorting, and other improvements. So the structure used in Vuex may change... not sure whether we should use something like `table.lookup` as we are doing with the Tree view structure, to quickly access the job & task (NB: the global `lookup` doesn't have the job anymore as it has been pruned in another delta-callback already).

Reply:
> the global lookup doesn't have the job anymore as it has been pruned in another delta-callback already

Does that mean the global flat data store doesn't contain jobs?

> something like table.lookup

Could we run the table off of the regular flat or tree stores?

Reply:
>Does that mean the global flat data store doesn't contain jobs?

Not anymore. It had the data, until the callback to apply the new `pruned` delta was applied to the global flat data.

I've been thinking about this issue, which happened with the tree view too.

Ideally I think we would want to

1. apply all the added & update in order (i.e. global, then each component like tree, then view, then graph, etc)
2. then apply the pruned deltas in reverse order (i.e. the components first, so that they can use the global lookup, and lastly the global callback to prune data)

That would solve the issue here I think (not sure about the tree view, since there we want to access the tree-node entry, not the global data). Will need to think more about this one.

>Could we run the table off of the regular flat or tree stores?

With some extra work, perhaps. The structure used for the table is

- task proxies
  * latest job (for now, @hjoliver I think mentioned that we want to display all jobs later, with some collapse/expand functionality)

The flat store has the data. But we would have to scan every entry whenever it's updated, and update a `computed` variable somewhere, with the structure for the component.

The tree store has the structure, but the task proxies are buried deeper in the tree, under cycle-points/family-proxies. So I would have to filter the tree store whenever it's updated, to re-build the data for the table (a `computed` property).